### PR TITLE
Updated awsProvider to allow manual specification of certificate auth…

### LINKED ISSUE #3958: "self signed certificate in certificate chain" - how to add self-signed cert to trusted store
Hi - Similar to #3230, I've just run into the same issue today. I'm using the latest version of Serverless, 1.17.0, and I know exactly what the issue is just not how to correct it.

I've recently begun testing deep packet inspection in my network. The firewall is decrypting, inspecting, and re-encrypting my SSL traffic for malware and such. There is a firewall generated, self-signed certificate deployed to the local CA trust store, and for other applications that don't use the OS's trust store I've had to add the self-signed cert somewhere/somehow.

For example, git requires:
`git config --global http.sslCAInfo <path to cert.pem>`

npm has a similar functionality for adding a reference to the cert or strict checking can be disabled with:
`npm config set strict-ssl false`

So... question is: how might I add or reference the self-signed cert in serverless?

```
Serverless: Packaging service...
  Serverless Error ---------------------------------------
  ServerlessError: self signed certificate in certificate chain
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless
  Your Environment Information -----------------------------
     OS:                     win32
     Node Version:           6.10.3
     Serverless Version:     1.17.0

```
In the serverless documentation for OpenWhisk, there is a reference to a property of the provider called "ignore_certs" that can be set to true. I'm using AWS, but tried this anyway and it did not change the result.
image
https://serverless.com/framework/docs/providers/openwhisk/guide/credentials/#register-with-openwhisk-platform-self-hosted

Any suggestions would be greatly appreciated.

Cheers!
Bryan

### PR DESCRIPTION
**Pull request by geofflancaster**
…orities.

## What did you implement: Allow manual specification of certificate authority for AWS provider.

Closes #3958 

## How did you implement it:

I followed a similar approach as the npm config (https://docs.npmjs.com/misc/config#ca) which allows users to specify a string of one or more certificate authority certificates(ca) or a string of one or more file paths to certificate authority certificates(cafile).

I also followed the AWS guidelines to define the httpsAgent and use the specified certificate authorities (http://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/node-registering-certs.html).

## How can we verify it:

I added tests cases. 

Before:
image

After: 
image

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by geofflancaster** [comment]
@pmuens Any updates on this?

File: lib/plugins/aws/provider/awsProvider.js [hunk 1]
```diff
@@ -116,6 +118,42 @@ class AwsProvider {
...
+      || process.env.HTTPS_CA
+      || process.env.https_ca;
+
+    const caCerts = [];
+
+    if (ca) {
+      // Can be a single certificate or multiple, comma separated.
+      const caArr = ca.split(',');
+      // Replace the newline -- https://stackoverflow.com/questions/30400341
+      caArr.forEach((cert) => {
```
Comment (RafalWilinski) [inline lib/plugins/aws/provider/awsProvider.js]: Replace `forEach` + `push` with:
```js
caCerts = caArr.map(cert => cert.replace(/\\n/g, '\n'));
```

File: lib/plugins/aws/provider/awsProvider.js [hunk 2]
```diff
@@ -116,6 +118,42 @@ class AwsProvider {
...
+      });
+    }
+
+    const cafile = process.env.cafile
+      || process.env.HTTPS_CAFILE
+      || process.env.https_cafile;
+
+    if (cafile) {
+      // Can be a single certificate file path or multiple paths, comma separated.
+      const caPathArr = cafile.split(',');
```
Comment (RafalWilinski) [inline lib/plugins/aws/provider/awsProvider.js]: Same here

Comment (geofflancaster) [reply lib/plugins/aws/provider/awsProvider.js]: Updated to use concat and map since the code change allows users to both specify a cert as a string and as a file simultaneously.

File: lib/classes/Utils.test.js [hunk 3]
```diff
@@ -671,6 +671,7 @@ describe('Utils', () => {
         provider: {
           name: 'google',
         },
+        package: {},
```
Comment (pmuens) [inline lib/classes/Utils.test.js]: Why is this change necessary?

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 4]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+      expect(typeof newAwsProvider.sdk.config.httpOptions.agent).to.not.equal('undefined');
+
+      // clear env
+      delete process.env.ca;
+    });
+
+    it('should set AWS cafile single', () => {
+      const tmpdir = os.tmpdir();
+      const certContents = '-----BEGIN CERTIFICATE-----\n-----END CERTIFICATE-----';
+      fs.writeFileSync(`${tmpdir}/ca.txt`, certContents);
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Could use `path.join` here to make this OS independent?

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 5]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
     });
 
+    it('should set AWS ca single', () => {
+      process.env.ca = '-----BEGIN CERTIFICATE-----\n-----END CERTIFICATE-----';
+      const newAwsProvider = new AwsProvider(serverless, options);
+
+      expect(typeof newAwsProvider.sdk.config.httpOptions.agent).to.not.equal('undefined');
+
+      // clear env
+      delete process.env.ca;
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: this should be done in an `afterEach` block.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 6]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+
+    it('should set AWS ca multiple', () => {
+      const certContents = '-----BEGIN CERTIFICATE-----\n-----END CERTIFICATE-----';
+      process.env.ca = `${certContents},${certContents}`;
+      const newAwsProvider = new AwsProvider(serverless, options);
+
+      expect(typeof newAwsProvider.sdk.config.httpOptions.agent).to.not.equal('undefined');
+
+      // clear env
+      delete process.env.ca;
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: This should be done in an `afterEach` block.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 7]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+      const tmpdir = os.tmpdir();
+      const certContents = '-----BEGIN CERTIFICATE-----\n-----END CERTIFICATE-----';
+      fs.writeFileSync(`${tmpdir}/ca.txt`, certContents);
+      process.env.cafile = `${tmpdir}/ca.txt`;
+      const newAwsProvider = new AwsProvider(serverless, options);
+
+      expect(typeof newAwsProvider.sdk.config.httpOptions.agent).to.not.equal('undefined');
+
+      // clear env
+      fs.unlink(`${tmpdir}/ca.txt`);
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Could use `path.join` here to make this OS independent?

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 8]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+      delete process.env.cafile;
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: This should be done in an `afterEach` block.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 9]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+    });
+
+    it('should set AWS cafile multiple', () => {
+      const tmpdir = os.tmpdir();
+      const certContents = '-----BEGIN CERTIFICATE-----\n-----END CERTIFICATE-----';
+      fs.writeFileSync(`${tmpdir}/ca1.txt`, certContents);
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Could use `path.join` here to make this OS independent?

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 10]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+      fs.writeFileSync(`${tmpdir}/ca2.txt`, certContents);
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Could use `path.join` here to make this OS independent?

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 11]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+      process.env.cafile = `${tmpdir}/ca1.txt,${tmpdir}/ca2.txt`;
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Could use `path.join` here to make this OS independent?

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 12]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+      const newAwsProvider = new AwsProvider(serverless, options);
+
+      expect(typeof newAwsProvider.sdk.config.httpOptions.agent).to.not.equal('undefined');
+
+      // clear env
+      fs.unlink(`${tmpdir}/ca1.txt`);
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: This should be done in an `afterEach` block.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 13]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+      fs.unlink(`${tmpdir}/ca2.txt`);
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: This should be done in an `afterEach` block.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 14]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+      delete process.env.cafile;
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: This should be done in an `afterEach` block.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 15]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+      const certContents = '-----BEGIN CERTIFICATE-----\n-----END CERTIFICATE-----';
+      fs.writeFileSync(`${tmpdir}/ca.txt`, certContents);
+      process.env.ca = certContents;
+      process.env.cafile = `${tmpdir}/ca.txt`;
+      const newAwsProvider = new AwsProvider(serverless, options);
+
+      expect(typeof newAwsProvider.sdk.config.httpOptions.agent).to.not.equal('undefined');
+
+      // clear env
+      fs.unlink(`${tmpdir}/ca.txt`);
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: This should be done in an `afterEach` block.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 16]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+      delete process.env.ca;
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: This should be done in an `afterEach` block.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 17]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+      delete process.env.cafile;
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: This should be done in an `afterEach` block.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 18]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+      // clear env
+      fs.unlink(`${tmpdir}/ca1.txt`);
+      fs.unlink(`${tmpdir}/ca2.txt`);
+      delete process.env.cafile;
+    });
+
+    it('should set AWS ca and cafile', () => {
+      const tmpdir = os.tmpdir();
+      const certContents = '-----BEGIN CERTIFICATE-----\n-----END CERTIFICATE-----';
+      fs.writeFileSync(`${tmpdir}/ca.txt`, certContents);
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Could use `path.join` here to make this OS independent?

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 19]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+      process.env.ca = certContents;
+      process.env.cafile = `${tmpdir}/ca.txt`;
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Could use `path.join` here to make this OS independent?

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 20]
```diff
@@ -51,6 +53,73 @@ describe('AwsProvider', () => {
...
+
+      // clear env
+      delete process.env.ca;
+    });
+
+    it('should set AWS cafile single', () => {
+      const tmpdir = os.tmpdir();
+      const certContents = '-----BEGIN CERTIFICATE-----\n-----END CERTIFICATE-----';
+      fs.writeFileSync(`${tmpdir}/ca.txt`, certContents);
+      process.env.cafile = `${tmpdir}/ca.txt`;
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Could use `path.join` here to make this OS independent?

File: lib/classes/Utils.test.js [hunk 21]
```diff
@@ -693,6 +694,7 @@ describe('Utils', () => {
         provider: {
           name: 'aws',
         },
+        package: {},
```
Comment (pmuens) [inline lib/classes/Utils.test.js]: Why is this change necessary?

File: lib/classes/Utils.test.js [hunk 22]
```diff
@@ -717,6 +719,7 @@ describe('Utils', () => {
...
```
Comment (pmuens) [inline lib/classes/Utils.test.js]: Why is this change necessary?

File: lib/classes/Utils.test.js (hunk 3 above)
Comment (geofflancaster) [reply lib/classes/Utils.test.js]: This was to fix a unit test bug as defined in #3600 that has since been merged. I'll revert and accept their changes.

File: lib/classes/Utils.test.js (hunk 21 above)
Comment (geofflancaster) [reply lib/classes/Utils.test.js]: This was to fix a unit test bug as defined in #3600 that has since been merged. I'll revert and accept their changes.

File: lib/plugins/aws/provider/awsProvider.test.js (hunk 5 above)
Comment (geofflancaster) [reply lib/plugins/aws/provider/awsProvider.test.js]: Would you like this done even though it's not applicable for all tests within this describe() block?

File: lib/plugins/aws/provider/awsProvider.test.js (hunk 12 above)
Comment (geofflancaster) [reply lib/plugins/aws/provider/awsProvider.test.js]: I'm new to Mocha so I might be wrong here but in my reading of the docs, this won't work.

Given the following:
- A hook should be defined in the describe block or at the root level. 
- fs.unlink() throws an exception if the file is not found
- not all tests in the constructor's describe() block require these files so the files will not exist

I don't think this will work. Thoughts?

File: lib/plugins/aws/provider/awsProvider.test.js (hunk 4 above)
Comment (geofflancaster) [reply lib/plugins/aws/provider/awsProvider.test.js]: complete.

File: lib/plugins/aws/provider/awsProvider.test.js (hunk 20 above)
Comment (geofflancaster) [reply lib/plugins/aws/provider/awsProvider.test.js]: complete.

File: lib/plugins/aws/provider/awsProvider.test.js (hunk 7 above)
Comment (geofflancaster) [reply lib/plugins/aws/provider/awsProvider.test.js]: complete.

File: lib/plugins/aws/provider/awsProvider.test.js (hunk 9 above)
Comment (geofflancaster) [reply lib/plugins/aws/provider/awsProvider.test.js]: complete.

File: lib/plugins/aws/provider/awsProvider.test.js (hunk 10 above)
Comment (geofflancaster) [reply lib/plugins/aws/provider/awsProvider.test.js]: complete.

File: lib/plugins/aws/provider/awsProvider.test.js (hunk 11 above)
Comment (geofflancaster) [reply lib/plugins/aws/provider/awsProvider.test.js]: complete.

File: lib/plugins/aws/provider/awsProvider.test.js (hunk 18 above)
Comment (geofflancaster) [reply lib/plugins/aws/provider/awsProvider.test.js]: complete.

File: lib/plugins/aws/provider/awsProvider.test.js (hunk 19 above)
Comment (geofflancaster) [reply lib/plugins/aws/provider/awsProvider.test.js]: complete.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 23]
```diff
@@ -46,9 +56,67 @@ describe('AwsProvider', () => {
...
+      const tmpdir = os.tmpdir();
+      const certContents = '-----BEGIN CERTIFICATE-----\n-----END CERTIFICATE-----';
+      const filePath = path.join(tmpdir, 'ca.txt');
+      fs.writeFileSync(filePath, certContents);
+      process.env.cafile = filePath;
+      const newAwsProvider = new AwsProvider(serverless, options);
+
+      expect(typeof newAwsProvider.sdk.config.httpOptions.agent).to.not.equal('undefined');
 
       // clear env
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: This cleanup should also be moved into the `afterEach` block.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 24]
```diff
@@ -46,9 +56,67 @@ describe('AwsProvider', () => {
...
+      const filePath1 = path.join(tmpdir, 'ca1.txt');
+      const filePath2 = path.join(tmpdir, 'ca2.txt');
+      fs.writeFileSync(filePath1, certContents);
+      fs.writeFileSync(filePath2, certContents);
+      process.env.cafile = `${filePath1},${filePath2}`;
+      const newAwsProvider = new AwsProvider(serverless, options);
+
+      expect(typeof newAwsProvider.sdk.config.httpOptions.agent).to.not.equal('undefined');
+
+      // clear env
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: This cleanup should also be moved into the `afterEach` block.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 25]
```diff
@@ -46,9 +56,67 @@ describe('AwsProvider', () => {
...
+      const certContents = '-----BEGIN CERTIFICATE-----\n-----END CERTIFICATE-----';
+      const filePath = path.join(tmpdir, 'ca.txt');
+      fs.writeFileSync(filePath, certContents);
+      process.env.ca = certContents;
+      process.env.cafile = filePath;
+      const newAwsProvider = new AwsProvider(serverless, options);
+
+      expect(typeof newAwsProvider.sdk.config.httpOptions.agent).to.not.equal('undefined');
+
+      // clear env
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: This cleanup should also be moved into the `afterEach` block.

File: lib/plugins/aws/provider/awsProvider.test.js (hunk 23 above)
Comment (geofflancaster) [reply lib/plugins/aws/provider/awsProvider.test.js]: "This" meaning the comment or the unlinkSync() call?

If the latter, I'm new to Mocha so I might be wrong here but in my reading of the docs, this won't work.

Given the following:

- A hook should be defined in the describe block or at the root level.
- fs.unlink() throws an exception if the file is not found
- not all tests in the constructor's describe() block require these files so for most tests the files will not exist and an exception will be thrown

Therefore, I don't think this will work. Thoughts?

Comment (geofflancaster) [reply lib/plugins/aws/provider/awsProvider.test.js]: @pmuens -- in case you didn't see this comment.

Comment (pmuens) [reply lib/plugins/aws/provider/awsProvider.test.js]: Thanks for the reply and pinging here @geofflancaster 👍

🤔 Good catch! Yes, I think you're right. The only feasible solution to solve this problem is to wrap all the tests which rely on the linking into a `describe()` block and then add a `afterEach()` block which takes care of the unlinking (via `fs.unlinkSync()`) there.

**Comment by geofflancaster** [comment]
@pmuens I have updated the test cases as you requested. For some reason travis-ci is throwing errors about "Too many requests" so several executions failed. This is caused by a changed that I merged in for test "#request() - should retry if error code is 429"

**Comment by shortjared** [comment]
Just had a client run into a need for this, great to see someone else has worked on solving it! Thanks @geofflancaster!

**Comment by ckabalan** [comment]
I'm the client @shortjared mentioned. Was able to get the .patch file for this PR and apply it directly to the `node_modules/serverless` directory and then it worked successfully.

1. Download the .patch by appending .patch to the Pull Request URL (after the number)
2. Move the .patch into the `node_modules/serverless` directory
3. Run the patch command: `patch -p1 < 4118.patch` (you'll have to decline patching the docs file)
4. Run with the environmental variable `cafile` set. Example: `cafile=/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem serverless deploy ...`

Obviously the cafile you point to will need to contain the custom certs your environment needs.

Would be great if this could be merged after appropriate testing. Thanks @geofflancaster.

**Comment by geofflancaster** [comment]
@pmuens This should be ready to merge. Let me know if you need any other changes.

**Comment by moraneva** [comment]
@pmuens any update on reviewing this? I am also facing this issue

**Comment by RafalWilinski** [comment]
I've tried using this feature and had no problems. Code also seems OK to me. @horike37, double check?

**Comment by moraneva** [comment]
@RafalWilinski @horike37 @pmuens

**Comment by geofflancaster** [comment]
Looks like the last accepted change was 26 days ago. Hopefully this project hasn't been abandoned but I do see that they have made comments on other issues recently.

Not sure why they're taking so long to respond to this one; especially now that it's been validated by others.

**Comment by horike37** [comment]
Thanks for merging @RafalWilinski :+1: 

>I've tried using this feature and had no problems. Code also seems OK to me. @horike37, double >check?

Sorry. I missed the comment :sweat_smile:

**Comment by geofflancaster** [comment]
@RafalWilinski thanks a ton for merging this! Looking forward to the next release!
